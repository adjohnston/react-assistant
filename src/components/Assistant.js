import React, { Component, Children, cloneElement } from 'react'
import PropTypes from 'prop-types'

const SpeechRecognition = window.SpeechRecognition || window.webkitSpeechRecognition

export default class Assistant extends Component {
  constructor(props) {
    super(props)

    this.state = {
      isListening: false
    }

    this.recognition = new SpeechRecognition()

    this.recognition.onend = () => {
      this.state.isListening
        ? this.recognition.start()
        : console.log('not listening');
    }

    this.recognition.onresult = ({ results }) => {
      const {
        confidence,
        transcript
      } = results[0][0]

      if (results[0].isFinal) {
        this.props.actions.forEach(action => {
          const identifier = Object.keys(action)[0]

          transcript.includes(identifier)
            ? action[identifier]()
            : null
        })
      }
    }
  }

  toggleListening = () => {
    this.setState(state => {
      return {
        isListening: !state.isListening
      }
    })
  }

  componentDidUpdate() {
    if (this.state.isListening) {
      this.recognition.start()
    } else {
      this.recognition.stop()
    }
  }

  render() {
    const isListening = this.state.isListening
    const children = this.props.children

    if (!SpeechRecognition)
      return null

    return Children.map(children, child => {
      switch (child.type.name) {
        case 'Switch':
          return cloneElement(child, {
            isListening,
            clickHandler: this.toggleListening
          })

        case 'StatusBar':
          return cloneElement(child, { isListening })

        default:
          return child
      }
    })
  }
}

Assistant.propTypes = {
  actions: PropTypes.arrayOf(
    PropTypes.object.isRequired
  ).isRequired
}
