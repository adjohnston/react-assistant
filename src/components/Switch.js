import React from 'react'
import PropTypes from 'prop-types'

const Switch = ({ children, isListening, actionHandler }) => {
  if (children)
    return children(isListening, actionHandler)

  return (
    <button
      onClick={actionHandler}>
      {isListening ? 'on' : 'off'}
    </button>
  )
}

Switch.propTypes = {
  isListening: PropTypes.bool,
  actionHandler: PropTypes.func,
  children: PropTypes.func
}

export default Switch
