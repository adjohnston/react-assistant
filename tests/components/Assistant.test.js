import React from 'react'
import renderer from 'react-test-renderer'
import Assistant from 'src/components/Assistant'
import Switch from 'src/components/Switch'

const mockProps = {
  actions: [{
    test: () => console.log('Just a test')
  }]
}

test('it should render null given no SpeechRecognition API', () => {
  const tree = renderer
    .create(<Assistant {...mockProps} />)
    .toJSON()

  expect(tree).toMatchSnapshot()
})

test('it should render null given no children', () => {
  window.SpeechRecognition = jest.fn()

  const tree = renderer
    .create(<Assistant {...mockProps} />)
    .toJSON()

  expect(tree).toMatchSnapshot()
})

test('it should render correctly with a default Switch component', () => {
  window.SpeechRecognition = jest.fn()

  const Component = (
    <Assistant {...mockProps}>
      <Switch />
    </Assistant>
  )

  const tree = renderer
    .create(Component)
    .toJSON()

  expect(tree).toMatchSnapshot()
})

test('it should render correctly with a custom Switch component', () => {
  window.SpeechRecognition = jest.fn()

  const Component = (
    <Assistant {...mockProps}>
      <Switch>
        {(isListening, actionHandler) => (
          <div
            onClick={actionHandler}>
            {isListening ? `I'm listening...` : `I'm not listening`}
          </div>
        )}
      </Switch>
    </Assistant>
  )

  const tree = renderer
    .create(Component)
    .toJSON()

  expect(tree).toMatchSnapshot()
})
