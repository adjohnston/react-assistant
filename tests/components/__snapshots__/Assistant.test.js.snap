// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it should render correctly with a custom Switch component 1`] = `
<div
  onClick={[Function]}
>
  I'm not listening
</div>
`;

exports[`it should render correctly with a default Switch component 1`] = `
<button
  onClick={[Function]}
>
  off
</button>
`;

exports[`it should render null given no SpeechRecognition API 1`] = `null`;

exports[`it should render null given no children 1`] = `null`;
